src/python/pants/backend/python/subsystems/pex_bootstrap/_ptex_launcher.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+"""Entrypoint script for a .ptex file generated with --generate-ptex.
+
+This script will build a .ipex file into a temporary directory, then execute it. If
+PANTS_PTEX_HYDRATE_ONLY_TO_IPEX=<filename> is set, however, the script will generate the ipex at
+<filename>, then exit without executing the .ipex file.
+"""
+
+import json
+import os
+import sys
+import tempfile
+
+from pex import resolver
+from pex.common import open_zip
+from pex.fetcher import Fetcher, PyPIFetcher
+from pex.interpreter import PythonInterpreter
+from pex.interpreter_constraints import matched_interpreters
+from pex.pex_builder import PEXBuilder
+from pex.pex_info import PexInfo
+
+
+APP_CODE_PREFIX = 'user_files/'
+
+
+def _strip_app_code_prefix(path):
+  if not path.startswith(APP_CODE_PREFIX):
+    raise ValueError(f"Path {path} in PTEX-INFO did not begin with '{APP_CODE_PREFIX}'.")
+  return path[len(APP_CODE_PREFIX):]
+
+
+HYDRATE_ONLY_NO_EXEC_ENV_VAR = 'PANTS_PTEX_HYDRATE_ONLY_TO_IPEX'
+
+
+def main(self):
+  filename_base, _ext = os.path.splitext(self)
+
+  # Build a .ipex file in a new temporary dir, each time.
+  td = tempfile.mkdtemp()
+  hydrated_ipex_file = os.environ.get(HYDRATE_ONLY_NO_EXEC_ENV_VAR, None)
+  oneshot_run = False
+  if not hydrated_ipex_file:
+    oneshot_run = True
+    hydrated_ipex_file = os.path.join(td, '{}.ipex'.format(filename_base))
+
+  sys.stderr.write('Hydrating {} to {}...\n'.format(self, hydrated_ipex_file))
+
+  with open_zip(self) as zf:
+    # Populate the pex with the pinned requirements and distribution names & hashes.
+    ipex_info = PexInfo.from_json(zf.read('IPEX-INFO'))
+    for interpreter in matched_interpreters(PythonInterpreter.all(),
+                                            ipex_info.interpreter_constraints):
+      ipex_builder = PEXBuilder(pex_info=ipex_info, interpreter=interpreter)
+      break
+    else:
+      raise ValueError(f'Could not resolve interpreter for constraints {ipex_info.interpreter_constraints}. '
+                       f'The IPEX-INFO for this .ptex was was:\n{ipex_info.dump(indent=4)}')
+
+    # Populate the pex with the needed code.
+    try:
+      ptex_info = json.loads(zf.read('PTEX-INFO').decode('utf-8'))
+      zf.extractall(td, ptex_info['code'])
+      for path in ptex_info['code']:
+        ipex_builder.add_source(os.path.join(td, path), env_filename=_strip_app_code_prefix(path))
+    except Exception as e:
+      raise ValueError(
+        f"Error: {e}. The PTEX-INFO for this .ptex file was:\n{json.dumps(ptex_info, indent=4)}"
+      ) from e
+
+  # Perform a fully pinned intransitive resolve to hydrate the install cache (not the
```

Comment:
Remove the "(not the pex!)" until it's actually true :) Maybe add a TODO here.

Reply:
Couldn't figure out what to write in the TODO! So left as is.